add `-status` flag to enable the Status service

gravityblast opened the pull request:
This PR adds a `-status` flag to enable the `Status` service.

gravityblast committed:
add `-status` flag to enable the Status service

dshulyak commented on geth/params/defaults.go:
```diff
@@ -33,7 +33,7 @@ const (
 	// APIModules is a list of modules to expose via any type of RPC (HTTP, IPC, in-proc)
 	// we also expose 2 limited personal APIs by overriding them in `api/backend.go`
-	APIModules = "eth,net,web3,shh,shhext"
+	APIModules = "eth,net,web3,shh,shhext,status"
```
thats bad idea, we dont want to allow `status` by default. i suggest to add `--api-modules` similarly to how geth is doing it

or finally add a config file support :)

gravityblast replied:
@dshulyak  If you start `statusd` without `-status`, the `Status` service is not visible. But I'm not sure about this part. Does it mean that by default that module is callable from dapps even if it's not public?

dshulyak replied:
as i remember we have now 2 bindings for rpc, public (node.Attach with public apis and this whitelist) and private (original node.Attach).

public rpc available to any dapp, so if status is in the whistelist... but if you say that Status is not enabled by default than it is not a big problem. but i still prefer to pass it explicitly

dshulyak reviewed (approved):
removing ❌

divan reviewed (approved):
LGTM
What's the use case to not enable status service, btw?

gravityblast commented:
@divan for the SDK https://github.com/status-im/status-go-sdk

<blockquote><img src="https://avatars3.githubusercontent.com/u/11767950?s=400&v=4" width="48" align="right"><div><img src="https://assets-cdn.github.com/favicon.ico" height="14"> GitHub</div><div><strong><a href="https://github.com/status-im/status-go-sdk">status-im/status-go-sdk</a></strong></div><div>Contribute to status-go-sdk development by creating an account on GitHub.</div></blockquote>

gravityblast committed:
remove status from default APIModules and add it only from statusd if specified

gravityblast commented:
@dshulyak @divan I removed the `status` service from the APIModules constant so that it's added to the config only from `cmd/statusd`

adambabik commented on cmd/statusd/main.go:
```diff
@@ -248,6 +249,11 @@ func makeNodeConfig() (*params.NodeConfig, error) {
 
+	nodeConfig.StatusServiceEnabled = *statusServiceEnabled
+	if nodeConfig.StatusServiceEnabled {
+		nodeConfig.AddAPIModule("status")
```
Why do we need to do it? If `status` API is private and the SDK will communicate via IPC, it should not be required.

`APIModules` config only affects what is exposed to DApps via HTTP (in our case `CallRPC` binding).

gravityblast replied:
@adambabik @adriacidre If `status` is hardcoded in the `APIModules`, this test is failing: https://github.com/status-im/status-go/blob/develop/t/e2e/services/status_api_test.go#L56

gravityblast replied:
so is it better to remove that test or better leave the kind of configuration?

adriacidre replied:
Basically `status` module is not whitelisted and not public, that makes it available only via IPC or inproc. It's being setup as a service [in here](https://github.com/status-im/status-go/pull/877/files#diff-32d99350ef62de1b6ea93cc22f7d66aaR173).

That test is checking exactly that those endpoints are not publicly accessible and in consequence accessible from any random dApp. On the other hand you have [private access tests](https://github.com/status-im/status-go/blob/develop/t/e2e/services/status_api_test.go#L59) to check they're accessible through IPC / inproc endpoints.

adambabik replied:
> That test is checking exactly that those endpoints are not publicly accessible and in consequence accessible from any random dApp.

"and in consequence **not** accessible from any random dApp" it should say, right?

gravityblast replied:
yeah that's why I add it from `statusd`. Even with IPC, if it's private is not accessible. If I leave it in the `APIModules` that test fails, if I don't add it from `statusd`, those methods are not available via HTTP/IPC

gravityblast replied:
In the past, I proposed to skip the signup/login process to avoid security problems in the SDK and leave the generation of the keys to the user or let bots use keys with `shh_generateKeyPair()` https://github.com/status-im/status-go/pull/877#issuecomment-385924009. but since we decided to support that service I made these changes

gravityblast replied:
Otherwise, we can use `nodeConfig.AddAPIModule("status")` only when `-http` is disabled, and `-ipc` is enabled. So that we ensure these API calls are possible only from local.

gravityblast committed:
remove AddAPIModule method

gravityblast commented:
ok I updated again the code. I removed the `AddAPIModule` that we don't need, and removed `status` from `APIModules` so that it's not available via HTTP, but it's available via IPC

dshulyak commented:
> ok I updated again the code. I removed the AddAPIModule that we don't need, and removed status from APIModules so that it's not available via HTTP, but it's available via IPC

i would :+1: for running status service over http, cause i usually run statusd in docker, so connecting via ipc would be quite a setback in terms of dev UX

adambabik commented:
I can see two options for that:
1. Provide a way to specify `APIModules` with a flag: `-apis=status` but that would require always putting the obvious ones like `eth` or `shh`,
2. Make `-status` to accept a list of interfaces: `-status=ipc` or `-status=ipc,http`.

gravityblast committed:
allow -status flag values to be http or ipc

gravityblast commented:
@adambabik @dshulyak I pushed a new commit.
With this one, I allow the `-status` flag to be `ipc` or `http` (if the service is public with `http`, it's automatically available on `ipc`).

`-status=ipc` requires `-ipc`, and `-status=http` requires `-http`. It's not done automatically so that the user needs to explicitly enable them.

I wrote some tests with the possible scenarios so that it can be descriptive in the future, and not broken with future refactoring.
